src/VSSetup.PowerShell/Lazy.cs

```diff
@@ -0,0 +1,76 @@
+﻿// <copyright file="Lazy.cs" company="Microsoft Corporation">
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt in the project root for license information.
+// </copyright>
+
+namespace Microsoft.VisualStudio.Setup
+{
+    using System;
+
+    /// <summary>
+    /// Creates an instance of <typeparamref name="T"/> only when <see cref="Value"/> is first accessed.
+    /// </summary>
+    /// <typeparam name="T">The type to create when <see cref="Value"/> is first accessed.</typeparam>
+    internal class Lazy<T> : IDisposable
+    {
+        private readonly Func<T> factory;
+        private bool hasValue = false;
+        private T value = default(T);
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Lazy{T}"/> class.
+        /// </summary>
+        /// <param name="factory">A <see cref="Func{TResult}"/> to create an instance of the type.</param>
+        public Lazy(Func<T> factory)
+        {
+            Validate.NotNull(factory, nameof(factory));
+            this.factory = factory;
+        }
+
+        /// <summary>
+        /// Gets a value indicating whether the <see cref="Value"/> has been created.
+        /// </summary>
+        public bool HasValue
+        {
+            get
+            {
+                lock (this)
```

Comment:
Suggest locking on a private object instead of `this` so that no one outside of class can affect this

Reply:
The class itself is `internal` and its a small project but, yes, does violate best practice regardless.